surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/TestStrategyApplier.java

```diff
@@ -44,32 +39,14 @@ public TestsToRun apply(List<String> orderStrategy) {
         return testsToRun;
     }
 
-    // TODO inverse creation of it - shouldn't belong to mvn package
     private TestExecutionPlanner getPlannerForStrategy(String orderStrategy) {
         final File projectDir = new File(System.getProperty("user.dir"));
-        final String previousCommit = System.getProperty("git.previous.commit", "HEAD");
-        final String commit = System.getProperty("git.commit", "HEAD");
-
         final String[] globPatternsAsArray = getGlobPatterns();
 
-        if (Objects.equals(orderStrategy, "new")) {
-            return new NewFilesDetector(projectDir, previousCommit, commit, globPatternsAsArray);
-        } else if (Objects.equals(orderStrategy, "changed")) {
-            return new ChangedFilesDetector(projectDir, previousCommit, commit, globPatternsAsArray);
-        } else if (Objects.equals(orderStrategy, "affected")) {
-            // TODO logic of inspecting git changes should be some where common so it is not recalculated several times
-            // TODO in fact there are at least two things to be put in a Context to be reused (git changes (main and test) and graph of dependencies between tests and main classes
-
-            // For now we recalculate everything
-            final Set<File> mainClasses =
-                new NewFilesDetector(projectDir, previousCommit, commit, "**/src/main/java/**/*.java").getFiles();
-
-            mainClasses.addAll(
-                new ChangedFilesDetector(projectDir, previousCommit, commit, "**/src/main/java/**/*.java").getFiles());
-
-            return new AffectedChangesDetector(new FileSystemTestClassDetector(projectDir), mainClasses);
-        } else if(Objects.equals(orderStrategy, "failed")) {
-            return new FailedTestsDetector();
+        for (final TestExecutionPlannerFactory testExecutionPlanner : ServiceLoader.load(TestExecutionPlannerFactory.class)) {
+            if (testExecutionPlanner.isFor(orderStrategy)) {
+                return testExecutionPlanner.create(projectDir, globPatternsAsArray);
```

Comment:
I am not sure if that should be there - my thinking is that the "client" (in this case surefire plugin) will decide when and how to instantiate it. At this point, we have a bit of coupling here too. Let's open an issue or document this concern in the wiki so that we don't forget to come back it.